mongo_connector/dest_mapping.py

```diff
@@ -0,0 +1,169 @@
+
+        for k, v in user_mapping.items():
+            self.set(k, v)
+
+    def set_plain(self, key, value):
+        """A utility function to set the corresponding plain variables"""
+        db, col = key.split(".", 1)
+        self.plain[key] = value
+        self.reverse_plain[value] = key
+        if col != "$cmd":
+            if db not in self.plain_db:
+                self.plain_db[db] = set([value.split(".")[0]])
+            else:
+                self.plain_db[db].add(value.split(".")[0])
+
+    def match(self, src, dst_pattern):
+        """If source string src matches dst, return the matchobject"""
+        reg_pattern = r'\A' + dst_pattern.replace('*', '(.*)') + r'\Z'
+        m = re.match(reg_pattern, src)
+        return m
+
+    def match_set(self, plain_src, dst_arr):
+        for x in dst_arr:
+            if plain_src == x:
+                return True
+            if "*" in x:
+                m = self.match(plain_src, x)
+                if m:
+                    return True
+
+        return False
+
+    def replace(self, src_match, map_pattern):
+        """Given the matchobject src_match,
+        replace corresponding '*' in map_pattern with src_match."""
+        wildcard_matched = src_match.group(1)
+        return map_pattern.replace("*", wildcard_matched)
+
+    def get(self, plain_src_ns, defaultval=""):
+        """Given a plain source namespace, return a mapped namespace if matched.
+        If no match and given defaultval, return it as a default value.
+        """
+        self.lock.acquire()
+        # search in plain mappings first
+        if plain_src_ns in self.plain.keys():
+            self.lock.release()
+            return self.plain[plain_src_ns]
+        else:
+            # search in wildcard mappings
+            # if matched, get a replaced mapped namespace
+            # and add to the plain mappings
+            for k, v in self.wildcard.items():
+                m_res = self.match(plain_src_ns, k)
+                if m_res:
+                    res = self.replace(m_res, v)
+                    if res is not None:
+                        if res not in self.reverse_plain:
+                            self.set_plain(plain_src_ns, res)
+                            self.lock.release()
+                            return res
+                        else:
+                            self.lock.release()
+                            raise errors.InvalidConfiguration(
+                                "Destination namespaces set wildcard pattern"
+                                " conflicts with the plain pattern.")
+
+        if defaultval:
+            self.lock.release()
+            return defaultval
+        else:
+            self.lock.release()
+            LOG.warn("Failed to find matched mapping for %s." % plain_src_ns)
+            return None
+
+    def set(self, key, value):
+        """Set the DestMapping instance to corresponding dictionary"""
+        self.lock.acquire()
```

Comment:
Acquire the lock using a with statement to simplify this code. See, https://docs.python.org/2/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement